File: backend/app/routers/plot.py
"""Plot-planning routes (ADR-0048 S4a/S4b/S5a/S7-Slice2; ADR-0053): plotlines,
cards, the board singleton, and templates.

The board, plotlines, cards, and templates are distinct sub-resources
(`/api/plot/board`, `/api/plot/plotlines/...`, `/api/plot/cards/...`,
`/api/plot/templates/...`) so no entry id can shadow another route. A plotline is
a plot-template instance (ADR-0053 §1): instantiating a template mints one, and an
ad-hoc plotline is just a plotline created with no beats — there is no separate
`/instances` resource.
"""
from __future__ import annotations

from fastapi import APIRouter

from app.models import (
    CardEntry,
    CardList,
    CreateCardRequest,
    CreatePlotlineRequest,
    CreatePlotTemplateRequest,
    PlotBoard,
    PlotBoardProjection,
    PlotContext,
    PlotlineEntry,
    PlotlineList,
    PlotTemplate,
    PlotTemplateList,
    RealizeCardRequest,
    SaveCardRequest,
    SavePlotBoardRequest,
    SavePlotlineRequest,
    SavePlotTemplateRequest,
)
from app.runtime import CurrentProject, translate_errors

router = APIRouter()


@router.get("/api/plot/board", response_model=PlotBoard)
def get_plot_board(project: CurrentProject) -> PlotBoard:
    """Open the plot board, creating it on first open (ADR-0048 §3)."""
    with translate_errors():
        return project.read_plot_board()


@router.put("/api/plot/board", response_model=PlotBoard)
def save_plot_board(project: CurrentProject, request: SavePlotBoardRequest) -> PlotBoard:
    with translate_errors():
        return project.save_plot_board(request)


@router.get("/api/plot/board/projection", response_model=PlotBoardProjection)
def get_plot_board_projection(project: CurrentProject) -> PlotBoardProjection:
    """The board's render model — plotlines + cards (with their refs) + the
    opaque layout, in one read (ADR-0048 S7a)."""
    with translate_errors():
        return project.read_plot_board_projection()


@router.get("/api/plot/board/context", response_model=PlotContext)
def get_plot_context(project: CurrentProject, as_of: str | None = None) -> PlotContext:
    """What the AI sees to reason about the plot (ADR-0048 S8a): the board's plot
    state — plotlines, arcs with full beat rosters, and cards with synopses / beat
    links / causal edges — spoiler-gated by manuscript reveal order.

    `as_of` (a card or scene id) anchors the gate: cards up to and including its
    reveal position are shown, later cards are withheld and only counted. Omit it
    for the whole board. This is the "what the AI sees" preview; the same read
    feeds Slice 8b's card brainstorm as prompt context."""
    with translate_errors():
        return project.read_plot_context(as_of)


@router.get("/api/plot/plotlines", response_model=PlotlineList)
def list_plotlines(project: CurrentProject) -> PlotlineList:
    with translate_errors():
        return project.list_plotlines()


@router.post("/api/plot/plotlines", response_model=PlotlineEntry)
def create_plotline(project: CurrentProject, request: CreatePlotlineRequest) -> PlotlineEntry:
    with translate_errors():
        return project.create_plotline(request)


@router.get("/api/plot/plotlines/{entry_id}", response_model=PlotlineEntry)
def get_plotline(project: CurrentProject, entry_id: str) -> PlotlineEntry:
    with translate_errors():
        return project.read_plotline(entry_id)


@router.put("/api/plot/plotlines/{entry_id}", response_model=PlotlineEntry)
def save_plotline(project: CurrentProject, entry_id: str, request: SavePlotlineRequest) -> PlotlineEntry:
    with translate_errors():
        return project.save_plotline(entry_id, request)


@router.delete("/api/plot/plotlines/{entry_id}", response_model=PlotlineList)
def delete_plotline(project: CurrentProject, entry_id: str) -> PlotlineList:
    with translate_errors():
        return project.delete_plotline(entry_id)


@router.get("/api/plot/cards", response_model=CardList)
def list_cards(project: CurrentProject) -> CardList:
    with translate_errors():
        return project.list_cards()


@router.post("/api/plot/cards", response_model=CardEntry)
def create_card(project: CurrentProject, request: CreateCardRequest) -> CardEntry:
    with translate_errors():
        return project.create_card(request)


@router.get("/api/plot/cards/{entry_id}", response_model=CardEntry)
def get_card(project: CurrentProject, entry_id: str) -> CardEntry:
    with translate_errors():
        return project.read_card(entry_id)


@router.put("/api/plot/cards/{entry_id}", response_model=CardEntry)
def save_card(project: CurrentProject, entry_id: str, request: SaveCardRequest) -> CardEntry:
    with translate_errors():
        return project.save_card(entry_id, request)


@router.delete("/api/plot/cards/{entry_id}", response_model=CardList)
def delete_card(project: CurrentProject, entry_id: str) -> CardList:
    with translate_errors():
        return project.delete_card(entry_id)


@router.post("/api/plot/cards/{entry_id}/realize", response_model=CardEntry)
def realize_card(project: CurrentProject, entry_id: str, request: RealizeCardRequest) -> CardEntry:
    """Create a scene from a card and attach it (ADR-0048 §1, *realize*)."""
    with translate_errors():
        return project.realize_card(entry_id, request)


@router.post("/api/plot/seed-from-manuscript", response_model=CardList)
def seed_cards_from_manuscript(project: CurrentProject) -> CardList:
    """Create one attached card per scene that has none (ADR-0048 §1/§S5)."""
    with translate_errors():
        return project.seed_cards_from_manuscript()


@router.get("/api/plot/templates", response_model=PlotTemplateList)
def list_plot_templates(project: CurrentProject) -> PlotTemplateList:
    with translate_errors():
        return project.list_plot_templates()


@router.post("/api/plot/templates", response_model=PlotTemplate)
def create_plot_template(project: CurrentProject, request: CreatePlotTemplateRequest) -> PlotTemplate:
    """Blank-create an owned `plot:template` (the non-fork path, #918)."""
    with translate_errors():
        return project.create_plot_template(request)


@router.get("/api/plot/templates/{entry_id}", response_model=PlotTemplate)
def get_plot_template(project: CurrentProject, entry_id: str) -> PlotTemplate:
    with translate_errors():
        return project.read_plot_template(entry_id)


@router.post("/api/plot/templates/{entry_id}/fork", response_model=PlotTemplate)
def fork_plot_template(project: CurrentProject, entry_id: str) -> PlotTemplate:
    """Clone an inherited Library / ancestor template into this project (ADR-0049 §5)."""
    with translate_errors():
        return project.fork_plot_template(entry_id)


@router.post("/api/plot/templates/{entry_id}/instantiate", response_model=PlotlineEntry)
def instantiate_plot_template(project: CurrentProject, entry_id: str) -> PlotlineEntry:
    """Apply a template to this book — snapshot its beats into a new, book-local,
    specializable plotline (ADR-0048 §3; ADR-0053 §2)."""
    with translate_errors():
        return project.instantiate_plot_template(entry_id)


@router.put("/api/plot/templates/{entry_id}", response_model=PlotTemplate)
def save_plot_template(project: CurrentProject, entry_id: str, request: SavePlotTemplateRequest) -> PlotTemplate:
    with translate_errors():
        return project.save_plot_template(entry_id, request)


@router.delete("/api/plot/templates/{entry_id}", response_model=PlotTemplateList)
def delete_plot_template(project: CurrentProject, entry_id: str) -> PlotTemplateList:
    with translate_errors():
        return project.delete_plot_template(entry_id)

File: frontend/src/lib/api/plot.ts
import type {
  PlotTemplate,
  PlotTemplateList,
  PlotBoardProjection,
  PlotBoard,
  PlotBoardLayout,
  CardEntry,
  CardList,
  PlotlineEntry,
  PlotlineList,
  CharacterArcEntry,
  CharacterArcList,
} from "@/lib/types";
import { request } from "./core";

export const plotApi = {
  // Plot templates (ADR-0048 S4c) — the ADR-0049 Library's second tenant. Same
  // browse/read/clone shape as prompts: list the resolved shelf, read one (with
  // its fail-closed `editable` verdict), clone an inherited one into an owned
  // editable copy, save/delete owned clones (inherited → 409 backend-side).
  listPlotTemplates() {
    return request<PlotTemplateList>("/plot/templates");
  },
  // Blank-create an owned template (#918) — the non-fork path. The backend defaults
  // a blank title, so callers may omit it and let the writer rename in the editor.
  createPlotTemplate(title = "") {
    return request<PlotTemplate>("/plot/templates", {
      method: "POST",
      body: JSON.stringify({ title }),
    });
  },
  getPlotTemplate(entryId: string) {
    return request<PlotTemplate>(`/plot/templates/${entryId}`);
  },
  forkPlotTemplate(entryId: string) {
    return request<PlotTemplate>(`/plot/templates/${entryId}/fork`, { method: "POST" });
  },
  savePlotTemplate(entry: PlotTemplate, body: string) {
    return request<PlotTemplate>(`/plot/templates/${entry.id}`, {
      method: "PUT",
      body: JSON.stringify({
        title: entry.title,
        body,
        template: entry.template,
        metadata: entry.metadata,
        base_revision: entry.revision,
      }),
    });
  },
  deletePlotTemplate(entryId: string) {
    return request<PlotTemplateList>(`/plot/templates/${entryId}`, { method: "DELETE" });
  },
  // The board's read model (ADR-0048 S7a): plotlines + cards (with their refs) +
  // the opaque layout, in one GET. Get-or-creates the `plot:board` singleton.
  getPlotBoardProjection() {
    return request<PlotBoardProjection>("/plot/board/projection");
  },
  // Persist the board layout (ADR-0048 S7c). PUT round-trips the opaque layout
  // dict with an optimistic base_revision; returns the board with its advanced
  // revision (the next save's base).
  savePlotBoard(payload: { base_revision: string; layout: PlotBoardLayout }) {
    return request<PlotBoard>("/plot/board", {
      method: "PUT",
      body: JSON.stringify(payload),
    });
  },
  // Plot cards + plotlines (ADR-0048 S5a/S5b) — the board's content ops, wired in
  // S7d. Cards and plotlines share the `plot/` folder + a book-local layered CRUD;
  // the endpoint path is the only family discriminator (the backend enforces an
  // is_a family guard on each). Attach/detach have no endpoint of their own — they
  // are a saveCard that sets / clears the `scene` ref in `metadata` (ADR §1).
  // The flat card list — the context picker's plot-card roster (ADR-0074 slice 6),
  // over which a plotline's selector expands to its current cards. The board still
  // reads its card set via the projection; this is the light list a picker needs.
  listCards() {
    return request<CardList>("/plot/cards");
  },
  // Create a single unattached card — the board's direct-authoring entry point
  // (#793), the per-card inverse of seed. Returns the created card so the caller can
  // open it to name it. No scene → it projects homeless until attached / realized.
  // `id` is supplied only by undo-of-delete / redo-of-create (ADR-0053 §7), to
  // restore a card under its original identity so other cards' causal_links
  // reconnect; a collision 409s. Omitted for a normal create (backend mints).
  createCard(title: string, id?: string) {
    return request<CardEntry>("/plot/cards", {
      method: "POST",
      body: JSON.stringify(id ? { title, id } : { title }),
    });
  },
  getCard(entryId: string) {
    return request<CardEntry>(`/plot/cards/${entryId}`);
  },
  saveCard(entry: CardEntry, body: string) {
    return request<CardEntry>(`/plot/cards/${entry.id}`, {
      method: "PUT",
      body: JSON.stringify({
        title: entry.title,
        body,
        metadata: entry.metadata,
        base_revision: entry.revision,
      }),
    });
  },
  deleteCard(entryId: string) {
    return request<CardList>(`/plot/cards/${entryId}`, { method: "DELETE" });
  },
  // Mint a scene from the card and attach it (ADR §1 *realize*). `parentId` places
  // the scene (null → the backend's first-container fallback). 409 if the card is
  // already attached (0..1 scene per card).
  realizeCard(entryId: string, parentId: string | null = null) {
    return request<CardEntry>(`/plot/cards/${entryId}/realize`, {
      method: "POST",
      body: JSON.stringify({ parent_id: parentId }),
    });
  },
  // Bulk inverse of realize (ADR §S5): one attached card per un-carded leaf scene,
  // in manuscript order. Idempotent — skips already-carded scenes.
  seedFromManuscript() {
    return request<CardList>("/plot/seed-from-manuscript", { method: "POST" });
  },
  // The plotline roster — the ReferencePicker's `plot` source (#742) and the
  // board's lanes both draw from this.
  listPlotlines() {
    return request<PlotlineList>("/plot/plotlines");
  },
  // Create an ad-hoc plotline (no template behind it) — the "New plotline" entry
  // point. Title only (colour + beats are authored afterward via savePlotline / the
  // board node); mirrors createCard. Returns the new node so the caller can place +
  // name it.
  // `id` is supplied only by undo/redo (ADR-0053 §7) to restore a plotline under
  // its original id so cards' beat_links + primary reconnect; a collision 409s.
  createPlotline(title: string, id?: string) {
    return request<PlotlineEntry>("/plot/plotlines", {
      method: "POST",
      body: JSON.stringify(id ? { title, id } : { title }),
    });
  },
  // Single plotline read/save/delete — the plotline document opener (#735): a
  // plotline backlink (a card's `plotline` ref) opens the thread in the editor to
  // rename / recolour / describe it. Book-local, so always editable (no Library
  // lock). Mirrors the card twins; delete returns the refreshed roster.
  getPlotline(entryId: string) {
    return request<PlotlineEntry>(`/plot/plotlines/${entryId}`);
  },
  savePlotline(entry: PlotlineEntry, body: string) {
    return request<PlotlineEntry>(`/plot/plotlines/${entry.id}`, {
      method: "PUT",
      body: JSON.stringify({
        title: entry.title,
        body,
        metadata: entry.metadata,
        base_revision: entry.revision,
      }),
    });
  },
  deletePlotline(entryId: string) {
    return request<PlotlineList>(`/plot/plotlines/${entryId}`, { method: "DELETE" });
  },
  // Character arcs (ADR-0080) — the plotline's sibling holder, same book-local flat
  // CRUD, distinct sub-resource (`/plot/character-arcs`, not `/plotlines`) per the
  // backend's family discriminator.
  listCharacterArcs() {
    return request<CharacterArcList>("/plot/character-arcs");
  },
  createCharacterArc(title: string, id?: string) {
    return request<CharacterArcEntry>("/plot/character-arcs", {
      method: "POST",
      body: JSON.stringify(id ? { title, id } : { title }),
    });
  },
  getCharacterArc(entryId: string) {
    return request<CharacterArcEntry>(`/plot/character-arcs/${entryId}`);
  },
  saveCharacterArc(entry: CharacterArcEntry, body: string) {
    return request<CharacterArcEntry>(`/plot/character-arcs/${entry.id}`, {
      method: "PUT",
      body: JSON.stringify({
        title: entry.title,
        body,
        metadata: entry.metadata,
        base_revision: entry.revision,
      }),
    });
  },
  deleteCharacterArc(entryId: string) {
    return request<CharacterArcList>(`/plot/character-arcs/${entryId}`, { method: "DELETE" });
  },
  // Snapshot a Library template's beats into a new owned plot:thread holder (ADR-0048
  // §3; ADR-0053 §1/§2; ADR-0080 §5 — a character-arc-family template yields a
  // plot:character_arc, any other family a plot:plotline). Lives among the template
  // routes backend-side; returns the created entry so the caller can place + edit it
  // on the board (the S3 palette's instantiate gesture) and — for the union — branch
  // on `entry_type` to route it to the right band. An ad-hoc plotline is a plain
  // createPlotline (no template behind it); there is no ad-hoc arc equivalent yet.
  instantiatePlotTemplate(templateId: string) {
    return request<PlotlineEntry | CharacterArcEntry>(`/plot/templates/${templateId}/instantiate`, { method: "POST" });
  },
};
